### PR TITLE
ICSD-1057 rogue server check

### PR DESCRIPTION
**Pull request by jzou-rbx**
Use Consul config retry_join list with hardcoded IP address (defined in Chef consul-cluster policies) as an allow list for consul servers.

Consul servers whose IP address that are not part of this allow list will ~~have its consul service stopped and~~ prevented it from starting.

For those already running as server but their ip address was removed from retry_join list. We do not shut them down immediately, these will be handled by an oncall.
For those who were not previously running as server (for example was shutdown or running as client), chef will restart consul  as part of the consul recipe, they will be blocked from starting.

https://roblox.atlassian.net/browse/ICSD-1057

Run book for overriding 
https://roblox.atlassian.net/wiki/spaces/IC/pages/2057044575/Service+Discovery+Runbook+New+Consul+Server+disallowed+from+joining+the+cluster

Please help review the followup PR https://github.com/Roblox/consul-cookbook/pull/45.

### CODE REVIEW

File: recipes/server_check.rb
Comment (sherrman) [inline recipes/server_check.rb]: Is there a `disable` for this also? A reboot of the machine will cause any enabled service units to start (which is the default state of the consul service.

File: recipes/server_check.rb
Comment (sherrman) [inline recipes/server_check.rb]: Does the conf file need user execute bit?

```suggestion
    mode '0644'
```

File: recipes/server_check_helper.rb
Comment (sherrman) [inline recipes/server_check_helper.rb]: Just a suggestion, take it or leave it. Maybe the name should be descriptive of what it's doing not the PMAI?

Like `is_server_disallowed?` or invert the logic with a different name.

**Comment by sherrman** [comment]
Accidentally removing an active server from this list could cause a chef run to stop a running server right?

Is there work to update the hashi-tools repo's consulctl to inspect the policy and check before applying it to a server not in the retry_join?

File: recipes/server_check.rb
Comment (jzou-rbx) [reply recipes/server_check.rb]: With this solution, simply rebooting the machine will not start the service. 

In order to bypass the systemd guardrail, need to remove the drop-in file at `/etc/systemd/system/consul.service.d/server_start_prevention.conf` .

In order to bypass the config file guardrail, need to overwrite `"server": true` in `/etc/consul/consul.json`

Then to allow server to listen on the desired DNS port, need to add 
SecureBits=keep-caps
AmbientCapabilities=CAP_NET_BIND_SERVICE
Capabilities=CAP_NET_BIND_SERVICE=+ep
CapabilityBoundingSet=CAP_SYSLOG CAP_NET_BIND_SERVICE

which the Chef cookbook would have done if `"server": true`.

Comment (jzou-rbx) [reply recipes/server_check.rb]: These extra guardrails ensure that the rogue server cant start up again after accidental restarts.

File: recipes/server_check.rb
Comment (jzou-rbx) [reply recipes/server_check.rb]: The user doesn't need to execute this, this is automatically picked up by systemd to check for condition before it starts a service.

Comment (glynnforrest) [reply recipes/server_check.rb]: > The user doesn't need to execute this

Exactly, so can we drop the +x bit? Change 7 to 6

File: test/helper/server_check_helper_test.rb
Comment (glynnforrest) [inline test/helper/server_check_helper_test.rb]: Personal preference - this is a double negative so a little hard to parse quickly.

```suggestion
        'allowed' => true,
```

File: recipes/server_check_helper.rb
Comment (glynnforrest) [inline recipes/server_check_helper.rb]: Personal preference, what do you think?

```suggestion
def is_server_allowed?(retry_join_list, my_ip)
```

Then we'd have

```ruby
if !is_server_allowed?
  # disable consul
```

It works better with `'allowed' => true` suggested below in tests.

File: test/helper/server_check_helper_test.rb
Comment (glynnforrest) [inline test/helper/server_check_helper_test.rb:6]: Looks like it'll work fine, but can we explicitly test for an AWS retry_join?

```ruby
retry_join: ["provider=aws tag_key=consul_cluster tag_value=cluster1"]
```

File: recipes/server_check.rb
Comment (glynnforrest) [inline recipes/server_check.rb]: Will this happen every time, even with no changes? Do we want that?

**Review by glynnforrest — COMMENTED** [review]
Looks great so far @jzou-rbx!

File: recipes/server_check.rb
Comment (jzou-rbx) [reply recipes/server_check.rb]: This will make systemd check for new configuration files every time, there is no impact for already running services per documentation. Currently I am just relying Chef resource's idempotency property for creating/deleting the conf file, but I can add an explicit file check to only load configuration if there is a change.

File: recipes/server_check_helper.rb
Comment (sherrman) [reply recipes/server_check_helper.rb]: Yeah, agree with the above point from @glynnforrest . A double negative is confusing.

**Comment by sherrman** [comment]
Thanks this is awesome!

Random thought... Did you think about failing the chef run after effectively disabling consul's systemd? This might be another way to signal failure.

But I don't know enough about how chef executes cookbooks to know if this is a good idea.

**Comment by fcuny-rbx** [comment]
Some general feedback:
- I don't think chef should stop consul if the IP address is not present
- I don't think we should add/remove drop-in file for systemd
- I'm not sure we should have the drop-in file check on itself when it's present, I think it might make more sense to call a script, in a ExecPreStart step
- If we go with a script, it should be the one checking the IP of the host against a list of expected IPs
- if the PreExecStart fails, we should write a metric under `/etc/grafana-agent/textfiles` saying that we have a rogue node, and an alert can be configured to alert an operator on what to do
  - it *might* be OK to fail nomad to start if the check fails, we should see what others think here

My main reasons here are:
- not letting chef stop consul: we had issues in the past where the chef server returned empty values for some configs, and this lead to bad outcome. I think I'd rather delegate (at this point) the operation to the oncall if an alert fires
- adding/removing drop-files has some edge cases, I'd rather have a stable interface, which is either the `ExecPreStart` in a drop-in file or the `ExecPreStart` in the main unit file, and not have to think about removing this drop-in (it's another step to remember to do). instead managing the list of IPs through chef is likely fine

**Comment by jzou-rbx** [comment]
> Thanks this is awesome!
> 
> Random thought... Did you think about failing the chef run after effectively disabling consul's systemd? This might be another way to signal failure.
> 
> But I don't know enough about how chef executes cookbooks to know if this is a good idea.



> Thanks this is awesome!
> 
> Random thought... Did you think about failing the chef run after effectively disabling consul's systemd? This might be another way to signal failure.
> 
> But I don't know enough about how chef executes cookbooks to know if this is a good idea.

Thanks Sherman, yeah initially I had a solution that uses systemd ExecStartPre which fails entire run, however it makes it not so straightward when running kitchen test etc. since that would also just error out.  I think not failing the run might be better because the Chef run in this case is performing the desired behavior of stopping Consul itself. We then can reserve Chef run failing to issues within Chef itself.

**Comment by jzou-rbx** [comment]
> Some general feedback:
> 
> * I don't think chef should stop consul if the IP address is not present
> * I don't think we should add/remove drop-in file for systemd
> * I'm not sure we should have the drop-in file check on itself when it's present, I think it might make more sense to call a script, in a ExecPreStart step
> * If we go with a script, it should be the one checking the IP of the host against a list of expected IPs
> * if the PreExecStart fails, we should write a metric under `/etc/grafana-agent/textfiles` saying that we have a rogue node, and an alert can be configured to alert an operator on what to do
>   
>   * it _might_ be OK to fail nomad to start if the check fails, we should see what others think here
> 
> My main reasons here are:
> 
> * not letting chef stop consul: we had issues in the past where the chef server returned empty values for some configs, and this lead to bad outcome. I think I'd rather delegate (at this point) the operation to the oncall if an alert fires
> * adding/removing drop-files has some edge cases, I'd rather have a stable interface, which is either the `ExecPreStart` in a drop-in file or the `ExecPreStart` in the main unit file, and not have to think about removing this drop-in (it's another step to remember to do). instead managing the list of IPs through chef is likely fine

Thank you @fcuny-rbx, just to clarify, do you mean not stop consul, but still block it from restarting/edit config,  or simply do not take any action just send metric to trigger alert and have someone manually perform these actions?  Also @glynnforrest and team any thoughts on this?

File: recipes/server_check_helper.rb
Comment (mechpen) [inline recipes/server_check_helper.rb]: why return true here?  should we disallow a node to become a server if the retry_join is invalid?

**Comment by fcuny-rbx** [comment]
> Thank you @fcuny-rbx, just to clarify, do you mean not stop consul, but still block it from restarting/edit config, or simply do not take any action just send metric to trigger alert and have someone manually perform these actions? Also @glynnforrest and team any thoughts on this?

I'm definitely in favor of not stopping consul in this case.

I'm split on the second part of your question. My instinct tells me that we should fail on restarting, if the host is not listed in the list of IPs. And we would want an alert when consul is running on a host that is not supposed to be in the cluster (and for that, we can have a systemd timer that run the same check as for `ExecPreStart` and generate the metric when it happens).

Definitely interested in other folks opinion, and hopefully I'm not derailing you from an agreed upon plan ...

**Comment by glynnforrest** [comment]
> Also @glynnforrest and team any thoughts on this?

Do we want to fail open or closed? If the chef/systemd protections fail for whatever reason, would we prefer rogue nodes accidentally added, or legitimate nodes accidentally removed?

File: recipes/server_check_helper.rb
Comment (jzou-rbx) [reply recipes/server_check_helper.rb]: Currently because we use Consul to block the server from starting,  so we probably only want to do this if there is an explicit retry list specified,  otherwise every server in that datacenter will be blocked.

Comment (mechpen) [reply recipes/server_check_helper.rb]: This recipe is only enabled when `node['consul']['config']['server'] === true`, right?
If a server has an invalid retry_join list, we should disable the server.
The retry_join list can be a DNS name, for example: `[ "consul.service.<dc>.consul"]`.  If a node has this config, it should not be allowed to be a consul server.

**Comment by mechpen** [comment]
> Do we want to fail open or closed? If the chef/systemd protections fail for whatever reason, would we prefer rogue nodes accidentally added, or legitimate nodes accidentally removed?

We are watching the cluster status during upgrades, so I am not too worried about some legitimate node getting silently removed.

File: recipes/server_check.rb
Comment (js3692) [reply recipes/server_check.rb]: +1 to `disable` - I think the above makes sense, but I think a disabled systemd service is a cleaner expression that this service should not be running, instead of a failed attempt to start on reboot

**Comment by jzou-rbx** [comment]
> > Do we want to fail open or closed? If the chef/systemd protections fail for whatever reason, would we prefer rogue nodes accidentally added, or legitimate nodes accidentally removed?
> 
> We are watching the cluster status during upgrades, so I am not too worried about some legitimate node getting silently removed.




> > Thank you @fcuny-rbx, just to clarify, do you mean not stop consul, but still block it from restarting/edit config, or simply do not take any action just send metric to trigger alert and have someone manually perform these actions? Also @glynnforrest and team any thoughts on this?
> 
> I'm definitely in favor of not stopping consul in this case.
> 
> I'm split on the second part of your question. My instinct tells me that we should fail on restarting, if the host is not listed in the list of IPs. And we would want an alert when consul is running on a host that is not supposed to be in the cluster (and for that, we can have a systemd timer that run the same check as for `ExecPreStart` and generate the metric when it happens).
> 
> Definitely interested in other folks opinion, and hopefully I'm not derailing you from an agreed upon plan ...

ExecPreStart cannot stop the service from starting (e.g. during restarts) unless it fails the entire Chef run, not sure if that is what we want though since it would also fail kitchen tests.  

I can remove the action that stops Consul server if it is already running,  and keep the double blockers (systemd start up blocking and config change `"server"=false"`) in place to prevent it from running on restarts, if we want to lean toward fail closed.

Alternatively, if we want to lean fail open, I can remove all actions and only rely on metrics. What do you guys think?

**Comment by jzou-rbx** [comment]
Validated in [test cluster] along with follow up PR https://github.com/Roblox/consul-cookbook/pull/45.

https://roblox.slack.com/archives/C04K9KKF38X/p1681862695418779

File: recipes/server_check_helper.rb
Comment (jzou-rbx) [reply recipes/server_check_helper.rb]: yes this recipe is only executed in default.rb when `node['consul']['config']['server'] === true`.
So the retry_join list we are checking is the retry_join list of the consul-cluster roles. 

I did not want to default to disallow when there is not a retry_join list defined:
1. Allow cases a cluster that maybe only has a single node and don't have a retry_join list defined yet.
2. There are cases where the consul-cluster role itself has a dynamic retry_join list, that we do not yet understand how to parse.

That is why I feel it is better if we just enable this guardrail where there is already an explict ip retry_join list defined in the consul-cluster policy role. Please let me know if this make sense, Thanks.

**Comment by jzou-rbx** [comment]
> > Thank you @fcuny-rbx, just to clarify, do you mean not stop consul, but still block it from restarting/edit config, or simply do not take any action just send metric to trigger alert and have someone manually perform these actions? Also @glynnforrest and team any thoughts on this?
> 
> I'm definitely in favor of not stopping consul in this case.
> 
> I'm split on the second part of your question. My instinct tells me that we should fail on restarting, if the host is not listed in the list of IPs. And we would want an alert when consul is running on a host that is not supposed to be in the cluster (and for that, we can have a systemd timer that run the same check as for `ExecPreStart` and generate the metric when it happens).
> 
> Definitely interested in other folks opinion, and hopefully I'm not derailing you from an agreed upon plan ...



> > Thank you @fcuny-rbx, just to clarify, do you mean not stop consul, but still block it from restarting/edit config, or simply do not take any action just send metric to trigger alert and have someone manually perform these actions? Also @glynnforrest and team any thoughts on this?
> 
> I'm definitely in favor of not stopping consul in this case.
> 
> I'm split on the second part of your question. My instinct tells me that we should fail on restarting, if the host is not listed in the list of IPs. And we would want an alert when consul is running on a host that is not supposed to be in the cluster (and for that, we can have a systemd timer that run the same check as for `ExecPreStart` and generate the metric when it happens).
> 
> Definitely interested in other folks opinion, and hopefully I'm not derailing you from an agreed upon plan ...

Hi @fcuny-rbx, just to update,  currently the new cookbook in this PR does not explicitly stop consul,  the consul service will fail on restarting (which when switching from a non consul server to a consul server, chef will restart consul).

The alerts are sent here (https://roblox.slack.com/archives/C04K9KKF38X/p1681862695418779)   from follow up PR https://github.com/Roblox/consul-cookbook/pull/45.  This guardrail can be manually disabled in a couple steps using https://roblox.atlassian.net/wiki/spaces/IC/pages/2057044575/Service+Discovery+Runbook+New+Consul+Server+disallowed+from+joining+the+cluster.  Is this ok?

**Comment by glynnforrest** [comment]
> default cookbook still restarts consul at the end of the run

nit - we reload consul, and don't restart it automatically with chef.

File: CHANGELOG.md
Comment (glynnforrest) [inline CHANGELOG.md:3]: Thanks for adding this 👍

File: recipes/server_check.rb
Comment (glynnforrest) [inline recipes/server_check.rb:34]: 👍

File: recipes/server_check_helper.rb
Comment (glynnforrest) [reply recipes/server_check_helper.rb]: > Allow cases a cluster that maybe only has a single node and don't have a retry_join list defined yet.

We don't do this anywhere. Always 5 or 3 servers.

> There are cases where the consul-cluster role itself has a dynamic retry_join list, that we do not yet understand how to parse.

This is handled in the conditional below with `is_valid_ip_address(x)?`, right? If not a list of IPs, it'll default to `true` at the end.

I'm on the fence, but side with @mechpen on this one. Perhaps an empty or missing retry_join should be false.

**Review by glynnforrest — APPROVED** [review]
Thanks a ton for this @jzou-rbx. I'm going to approve with the following caveats:

* Maybe reconsider `true` vs `false` for an empty retry_join
* Resolve the outstanding questions from @mechpen and @fcuny-rbx
* We've tested the servers and a single rogue node in [redacted test cluster]. Can we also test all the clients now? That'll be another policyfiles change.
* Get approval from a couple of others on the team.

File: recipes/server_check_helper.rb
Comment (jzou-rbx) [reply recipes/server_check_helper.rb]: > > Allow cases a cluster that maybe only has a single node and don't have a retry_join list defined yet.
> 
> We don't do this anywhere. Always 5 or 3 servers.  
> 
> > There are cases where the consul-cluster role itself has a dynamic retry_join list, that we do not yet understand how to parse.
> 
> This is handled in the conditional below with `is_valid_ip_address(x)?`, right? If not a list of IPs, it'll default to `true` at the end.
> 
> I'm on the fence, but side with @mechpen on this one. Perhaps an empty or missing retry_join should be false.

I see, what I meant was that currently the logic is to default to allow in these cases:
1. nil or empty retry_join   ( I will change this to disallow per comments).
2. dynamic list (will keep defaulting to allow).

File: recipes/server_check.rb
Comment (glynnforrest) [reply recipes/server_check.rb]: Resolved in https://github.com/Roblox/consul-cookbook/pull/44/commits/a3210fa5d0d7047e244e6446f47c90bc1ccd1a90

File: recipes/server_check.rb
Comment (glynnforrest) [reply recipes/server_check.rb]: Fixed in https://github.com/Roblox/consul-cookbook/pull/44/commits/a3210fa5d0d7047e244e6446f47c90bc1ccd1a90

File: recipes/server_check_helper.rb
Comment (glynnforrest) [reply recipes/server_check_helper.rb]: Fixed in Fixed in https://github.com/Roblox/consul-cookbook/pull/44/commits/a3210fa5d0d7047e244e6446f47c90bc1ccd1a90

File: recipes/server_check.rb
Comment (sherrman) [inline recipes/server_check.rb]: Directories should have the execution bit

```suggestion
  mode '0755'
```

**Review by sherrman — COMMENTED** [review]
I think whatever the effect of the chef run is. We want the overall system to be safe (and ideally, ideally the state is the same):

1. After chef-client is done.
2. The "rogue" node is restarted.

What are the situations this is intended to protect against? And what is it not?

For me understanding that matters for if we should stop consul or not. The rogue node problem has a lot of facets and not all of the things we're doing address all facets.

**Comment by jzou-rbx** [comment]
> I think whatever the effect of the chef run is. We want the overall system to be safe (and ideally, ideally the state is the same):
> 
> 1. After chef-client is done.
> 2. The "rogue" node is restarted.
> 
> What are the situations this is intended to protect against? And what is it not?
> 
> For me understanding that matters for if we should stop consul or not. The rogue node problem has a lot of facets and not all of the things we're doing address all facets.

Initially the PR stops the consul service explicitly when classified as rogue server, this would stop the consul server immediately even if it is currently running.   Then from Franck's comment above, I changed the PR to only block consul service from restarting, this: 
1. For those already running as server but their ip address was removed from retry_join list. We do not shut them down immediately, the alert will fire and we let an oncall person handle it.
2. For those who were not previously running as server (for example was shutdown or running as client),  chef will restart consul anyways as part of the consul recipe, they will be blocked from starting.

**Review by sherrman — APPROVED** [review]
> Thanks Sherman, yeah initially I had a solution that uses systemd ExecStartPre which fails entire run, however it makes it not so straightward when running kitchen test etc. since that would also just error out. I think not failing the run might be better because the Chef run in this case is performing the desired behavior of stopping Consul itself. We then can reserve Chef run failing to issues within Chef itself.

This came up in a response to @fcuny-rbx also. How does `ExecStartPre` fail chef? It's just a command run during systemd unit's run. If any of the `ExecStartPre` lines fail, the systemd unit is failed and won't continue running the remaining commands I think.

If I'm understanding how `ExecStartPre` works, I personally think a script in `ExecStartPre` is cleaner. It decouples the starting of consul.service from the chef run. Using a file marker requires chef to do the work.

But if that's the biggest gain, we should err on merging what's done. Agree, we should hash it out with @fcuny-rbx before merging.

**Comment by jzou-rbx** [comment]
> > Thanks Sherman, yeah initially I had a solution that uses systemd ExecStartPre which fails entire run, however it makes it not so straightward when running kitchen test etc. since that would also just error out. I think not failing the run might be better because the Chef run in this case is performing the desired behavior of stopping Consul itself. We then can reserve Chef run failing to issues within Chef itself.
> 
> This came up in a response to @fcuny-rbx also. How does `ExecStartPre` fail chef? It's just a command run during systemd unit's run. If any of the `ExecStartPre` lines fail, the systemd unit is failed and won't continue running the remaining commands I think.
> 
> If I'm understanding how `ExecStartPre` works, I personally think a script in `ExecStartPre` is cleaner. It decouples the starting of consul.service from the chef run. Using a file marker requires chef to do the work.
> 
> But if that's the biggest gain, we should err on merging what's done. Agree, we should hash it out with @fcuny-rbx before merging.

Just to clarify, my point was the ExecStartPre cannot block service from starting, unless it errors out which not just kills the service start, but the entire chef pipeline.

A possible alternative would be to ExecStartPre run a script to create the fileamarker if classifies as rogue node. Although we would still depend on Chef run to set this up in the first place, and we would need to move the logic outside of the Chef ecosystem (currently I tried to use Chef to classify rogue node, create/delete files, update metrics etc.), would this be worth considering?

**Comment by sherrman** [comment]
> Just to clarify, my point was the ExecStartPre cannot block service from starting, unless it errors out which not just kills the service start, but the entire chef pipeline.

Ah I see. OK according to the man page, ExecStartPre puts the service into a "failed" state. But using Conditions (there's a huge number of them) is a "mostly silent" skipping and does not mark them failed.

Do you think this is worth explaining as a comment in the cookbook?
 
> A possible alternative would be to ExecStartPre run a script to create the fileamarker if classifies as rogue node. Although we would still depend on Chef run to set this up in the first place, and we would need to move the logic outside of the Chef ecosystem (currently I tried to use Chef to classify rogue node, create/delete files, update metrics etc.), would this be worth considering?

For me no. This starts getting into complex/convoluted territory.

**Review by mechpen — APPROVED** [review]

File: .kitchen.yml
Comment (fcuny-rbx) [inline .kitchen.yml]: I think `is_decommissioned_server` might be a better name. I don't think new comers will understand what we mean by rogue.

File: recipes/server_check.rb
Comment (fcuny-rbx) [inline recipes/server_check.rb:46]: It still feels weird to me that we are checking the drop-in file for this check, but I don't think it's wrong. I would personally prefer to delegate this logic to a script and rely on `ExecStartPre` and a status code. However, I understand this adds more complexity and we may want to go with this already tested and validated approach.

File: recipes/server_check.rb
Comment (fcuny-rbx) [inline recipes/server_check.rb:20]: `retry_join` seems to be an already existing attribute, however I could not find any documentation about it in this cookbook. It might be worth updating the README and document what it is, it would make it easier to understand this in the future.

**Review by fcuny-rbx — APPROVED** [review]
Based on the testing / validation done (I read the slack thread linked above), and the feedback from the team, let's go ahead with this change.

I've left a comment about naming (`is_rogue_node` vs. `is_decommissioned_server`), documenting `retry_join`, but these are nits.

File: .kitchen.yml
Comment (jzou-rbx) [reply .kitchen.yml]: Got it, instead of is_rogue_server, I renamed to is_unexpected_server to be even more general.

File: recipes/server_check.rb
Comment (jzou-rbx) [reply recipes/server_check.rb:20]: Added some information about retry_join in the comments.

**Pull request merged**
